File: app/src/main/java/me/sunzheng/mana/home/mybangumi/FavoritesFragment.java
package me.sunzheng.mana.home.mybangumi;

import android.content.Context;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.DefaultItemAnimator;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;
import me.sunzheng.mana.R;
import me.sunzheng.mana.home.HomeContract;

/**
 * Created by Sun on 2017/6/22.
 */

public class FavoritesFragment extends Fragment implements HomeContract.MyBangumi.View {
    RecyclerView mRecyclerView;
    SwipeRefreshLayout mSwipeRefreshLayout;
    HomeContract.MyBangumi.Presenter mPresenter;

    public FavoritesFragment() {
    }

    public static FavoritesFragment newInstance() {
        FavoritesFragment fragment = new FavoritesFragment();
        return fragment;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
    }

    @Override
    public void onResume() {
        super.onResume();
    }

    @Override
    public void onPause() {
        super.onPause();
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_myfavorites, container, false);
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        mRecyclerView = view.findViewById(R.id.recycler_view);
        mSwipeRefreshLayout = view.findViewById(R.id.swiperefreshlayout);
        mSwipeRefreshLayout.setColorSchemeResources(R.color.colorPrimary);
        mSwipeRefreshLayout.setEnabled(false);
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        if (mPresenter != null)
            mPresenter.subscribe();
    }

    @Override
    public void onDetach() {
        super.onDetach();
        if (mPresenter != null)
            mPresenter.unsubscribe();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
    }

    @Override
    public void setPresenter(HomeContract.MyBangumi.Presenter presenter) {
        mPresenter = presenter;
    }

    @Override
    public void showProgressIntractor(boolean active) {
        if (mSwipeRefreshLayout == null)
            return;
        mSwipeRefreshLayout.setRefreshing(active);
    }

    @Override
    public void showEmpty() {

    }

    @Override
    public void onFilter(int status) {
        mPresenter.setFilter(status);
    }

    @Override
    public void setAdapter(RecyclerView.Adapter adapter) {
        if (mRecyclerView == null)
            return;
        if (mRecyclerView.getAdapter() == null) {
            mRecyclerView.setAdapter(adapter);
            mRecyclerView.setItemAnimator(new DefaultItemAnimator());
        } else
            mRecyclerView.swapAdapter(adapter, true);
        if (mRecyclerView.getLayoutManager() == null)
            mRecyclerView.setLayoutManager(new LinearLayoutManager(getActivity()));
    }
}

File: app/src/main/java/me/sunzheng/mana/home/mybangumi/FavoritesFragment.kt
package me.sunzheng.mana.home.mybangumi

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.lifecycle.LiveData
import androidx.lifecycle.ViewModel
import androidx.recyclerview.widget.DefaultItemAnimator
import androidx.recyclerview.widget.DividerItemDecoration
import com.google.gson.Gson
import dagger.hilt.android.AndroidEntryPoint
import dagger.hilt.android.lifecycle.HiltViewModel
import me.sunzheng.mana.BangumiDetailsActivity
import me.sunzheng.mana.R
import me.sunzheng.mana.core.net.ApiResponse
import me.sunzheng.mana.core.net.v2.ApiService
import me.sunzheng.mana.core.net.v2.NetworkBoundResource
import me.sunzheng.mana.core.net.v2.database.BangumiDao
import me.sunzheng.mana.core.net.v2.database.BangumiEntity
import me.sunzheng.mana.core.net.v2.database.FavirouteDao
import me.sunzheng.mana.core.net.v2.database.FavriouteEntity
import me.sunzheng.mana.databinding.FragmentMyfavoritesBinding
import me.sunzheng.mana.home.mybangumi.wrapper.FavoriteWrapper
import me.sunzheng.mana.home.onair.OnAirItemRecyclerViewAdapter
import java.util.*
import javax.inject.Inject
import javax.inject.Named

/**
 * Created by Sun on 2017/6/22.
 */
@AndroidEntryPoint
class FavoritesFragment @Inject constructor() : Fragment() {
    companion object {
        @JvmStatic
        val ARGS_DATA_PARCEL_ARRAY = "${FavoritesFragment::class.java.simpleName}_data"

        @JvmStatic
        fun newInstance(list: List<BangumiEntity>) = FavoritesFragment().apply {
            arguments = Bundle().apply {
                putParcelableArrayList(
                    ARGS_DATA_PARCEL_ARRAY,
                    ArrayList<BangumiEntity>(list)
                )
            }
        }
    }

    lateinit var binding: FragmentMyfavoritesBinding
    lateinit var args: Bundle
    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        binding = FragmentMyfavoritesBinding.inflate(inflater)
        binding.lifecycleOwner = this
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        args = savedInstanceState?.getBundle("args") ?: requireArguments()
        args.getParcelableArrayList<BangumiEntity>(ARGS_DATA_PARCEL_ARRAY)?.let {
            OnAirItemRecyclerViewAdapter(it) { v, _, _, m ->
                if (m is BangumiEntity)
                    BangumiDetailsActivity.newInstance(
                        requireActivity(),
                        m,
                        v.findViewById(R.id.item_album)
                    )
            }
        }?.run {
            binding.recyclerView.adapter = this
            binding.recyclerView.itemAnimator = DefaultItemAnimator()
            binding.recyclerView.addItemDecoration(
                DividerItemDecoration(
                    requireContext(),
                    DividerItemDecoration.VERTICAL
                )
            )
        }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        outState.putBundle("args", args)
        super.onSaveInstanceState(outState)
    }
}

@HiltViewModel
class MyFavoriteViewModel @Inject constructor() : ViewModel() {
    @Inject
    lateinit var favriouteDao: FavirouteDao

    @Inject
    lateinit var bangumiDao: BangumiDao

    @Inject
    lateinit var apiService: ApiService

    @Named("userName")
    @Inject
    lateinit var userName: String
    val repository: FavoritesRepository by lazy {
        FavoritesRepository().also {
            it.favriouteDao = favriouteDao
            it.bangumiDao = bangumiDao
            it.apiService = apiService
            it.userName = userName
        }
    }

    fun filter(status: Int = 0, page: Int = 0) = repository.query(status, page)
}

class FavoritesRepository {
    lateinit var favriouteDao: FavirouteDao
    lateinit var bangumiDao: BangumiDao
    lateinit var apiService: ApiService
    lateinit var userName: String
    fun query(status: Int, page: Int) =
        object : NetworkBoundResource<List<BangumiEntity>, FavoriteWrapper>() {
            override fun saveCallResult(item: FavoriteWrapper) {
                item.data?.forEach {
                    bangumiDao.insert(Gson().fromJson(Gson().toJson(it), BangumiEntity::class.java))
                }
                item.data.map {
                    FavriouteEntity(
                        UUID.fromString(it.id),
                        status = it.favoriteStatus,
                        userName = userName,
                        it.unwatched_count
                    )
                }.forEach {
                    favriouteDao.insert(it)
                }
            }

            override fun shouldFetch(data: List<BangumiEntity>?): Boolean = true

            override fun loadFromDb(): LiveData<List<BangumiEntity>> =
                favriouteDao.queryBangumiList(status, userName)

            override fun createCall(): LiveData<ApiResponse<FavoriteWrapper>> =
                apiService.listMyBangumi(status)
        }.asLiveData()
}